sys/uio.c

```diff
@@ -0,0 +1,88 @@
+/*
+ * Phoenix-RTOS
+ *
+ * libphoenix
+ *
+ * sys/uio.c
+ *
+ * Copyright 2021 Phoenix Systems
+ * Author: Ziemowit Leszczynski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <unistd.h>
+#include <limits.h>
+#include <sys/uio.h>
+
+
+static int checkv(const struct iovec *iov, int iovcnt)
+{
+	ssize_t max_len = SSIZE_MAX;
+
+	if (iovcnt <= 0 || iovcnt > IOV_MAX)
+		return -1;
+
+	for (unsigned int i = 0; i < iovcnt; ++i) {
+		if (iov[i].iov_len > max_len)
+			return -1;
+		max_len -= iov[i].iov_len;
+	}
+
+	return 0;
+}
+
+
+ssize_t readv(int fd, const struct iovec *iov, int iovcnt)
+{
+	ssize_t tot_len = 0;
+
+	if (-1 == checkv(iov, iovcnt))
+		return -EINVAL;
+
+	/* TODO: file locking */
+
+	for (unsigned int i = 0; i < iovcnt; ++i) {
+		if (iov[i].iov_len > 0) {
+			ssize_t n = read(fd, iov[i].iov_base, iov[i].iov_len);
+			if (n == -1)
+				return -1;
+
+			tot_len += n;
+
+			if (n < iov[i].iov_len)
+				break;
+		}
+	}
+
+	return tot_len;
+}
+
+
+ssize_t writev(int fd, const struct iovec *iov, int iovcnt)
+{
+	ssize_t tot_len = 0;
+
+	if (-1 == checkv(iov, iovcnt))
+		return -EINVAL;
+
+	/* TODO: file locking */
+
+	for (unsigned int i = 0; i < iovcnt; ++i) {
+		if (iov[i].iov_len > 0) {
+			ssize_t n = write(fd, iov[i].iov_base, iov[i].iov_len);
+			if (n == -1)
+				return -1;
+
+			tot_len += n;
+
+			if (n < iov[i].iov_len)
+				break;
```

Comment:
Why we give up if we wrote less than iov_len? I think we should write the missing data from current buffer, not move on to the next buffer

Reply:
Check previous reply.